src/stan/mcmc/hmc/hamiltonians/softabs_metric.hpp

```diff
@@ -0,0 +1,213 @@
+
+        Eigen::VectorXd a(z.p.size());
+
+        for (idx_t n = 0; n < z.p.size(); ++n)
+          a(n) = sqrt(z.softabs_lambda(n)) * rand_unit_gaus();
+
+        z.p = z.eigen_deco.eigenvectors() * a;
+      }
+
+      void init(
+        softabs_point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) {
+        update_metric(z, info_writer, error_writer);
+        update_metric_gradient(z, info_writer, error_writer);
+      }
+
+      void update_metric(
+        softabs_point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) {
+        // Compute the Hessian
+        stan::math::hessian<double>(
+          softabs_fun<Model>(this->model_, 0), z.q, z.V, z.g, z.hessian);
+
+        z.V *= -1;
+        z.g *= -1;
+        z.hessian *= -1;
+
+        // Compute the eigen decomposition of the Hessian,
+        // then perform the SoftAbs transformation
+        z.eigen_deco.compute(z.hessian);
+
+        for (idx_t i = 0; i < z.q.size(); ++i) {
+          double lambda = z.eigen_deco.eigenvalues()(i);
+          double alpha_lambda = z.alpha * lambda;
+
+          double softabs_lambda = 0;
+
+          // Thresholds defined such that the approximation
+          // error is on the same order of double precision
+          if (std::fabs(alpha_lambda) < 1e-4) {
+           softabs_lambda = (1.0
+                             + (1.0 / 3.0) * alpha_lambda * alpha_lambda)
+                             / z.alpha;
+          } else if (std::fabs(alpha_lambda) > 18) {
+           softabs_lambda = std::fabs(lambda);
+          } else {
+           softabs_lambda = lambda / std::tanh(alpha_lambda);
+          }
+
+          z.softabs_lambda(i) = softabs_lambda;
+          z.softabs_lambda_inv(i) = 1.0 / softabs_lambda;
+        }
+
+        // Compute the log determinant of the metric
+        z.log_det_metric = 0;
+        for (idx_t i = 0; i < z.q.size(); ++i)
+         z.log_det_metric += std::log(z.softabs_lambda(i));
+      }
+
+      void update_metric_gradient(
+        softabs_point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) {
+        // Compute the pseudo-Jacobian of the SoftAbs transform
+        for (idx_t i = 0; i < z.q.size(); ++i) {
+          for (idx_t j = 0; j <= i; ++j) {
+            double delta =   z.eigen_deco.eigenvalues()(i)
+                           - z.eigen_deco.eigenvalues()(j);
+
+            if (std::fabs(delta) < 1e-10) {
+              double lambda = z.eigen_deco.eigenvalues()(i);
+              double alpha_lambda = z.alpha * lambda;
+
+              // Thresholds defined such that the approximation
+              // error is on the same order of double precision
+              if (std::fabs(alpha_lambda) < 1e-4) {
```

Comment:
(fix now)  This 1e-4 threshold appears all over, so should be defined as a static const somewhere.